### PR TITLE
Mongo Source CSFLE support and Cosmos NoSQL Always Encrypted support

### PR DESCRIPTION
**Pull request by fonsecamar**
- Adding support for MongoDB Client Side Field Level Encryption: auto decrypt in Mongo source
- Adding support for Cosmos NoSQL Always Encrypted
- Adding basic transformations (destination field name and type). Can be useful to convert _id to id and non-string shard key to Cosmos NoSQL.
- Updating some dependencies due to existing vulnerability

### CODE REVIEW

File: Extensions/Cosmos/Cosmos.DataTransfer.CosmosExtension/CosmosExtensionServices.cs [hunk 1]
```diff
@@ -36,7 +38,17 @@ public static CosmosClient CreateClient(CosmosSettingsBase settings, string disp
             if (settings.UseRbacAuth)
             {
                 TokenCredential tokenCredential = new DefaultAzureCredential(includeInteractiveCredentials: settings.EnableInteractiveCredentials);
-                cosmosClient = new CosmosClient(settings.AccountEndpoint, tokenCredential, clientOptions);
+
+                if(settings.InitClientEncryption)
+                {
+                    var keyResolver = new KeyResolver(tokenCredential);
+                    cosmosClient = new CosmosClient(settings.AccountEndpoint, tokenCredential, clientOptions)
+                        .WithEncryption(keyResolver, KeyEncryptionKeyResolverName.AzureKeyVault);
```
Comment (markjbrown) [inline Extensions/Cosmos/Cosmos.DataTransfer.CosmosExtension/CosmosExtensionServices.cs:46]: The Readme for the cosmos extension is missing the key vault information for the client-side encryption information needed here. Can you add?

File: Extensions/Mongo/Cosmos.DataTransfer.MongoExtension/MongoDataSourceExtension.cs [hunk 2]
```diff
@@ -19,7 +19,7 @@ public async IAsyncEnumerable<IDataItem> ReadAsync(IConfiguration config, ILogge
 
         if (!string.IsNullOrEmpty(settings.ConnectionString) && !string.IsNullOrEmpty(settings.DatabaseName))
         {
-            var context = new Context(settings.ConnectionString, settings.DatabaseName);
+            var context = new Context(settings.ConnectionString, settings.DatabaseName, settings.KeyVaultNamespace, settings.KMSProviders);
```
Comment (markjbrown) [inline Extensions/Mongo/Cosmos.DataTransfer.MongoExtension/MongoDataSourceExtension.cs:22]: Can you fix this async warning on line 38?

**Review by markjbrown — COMMENTED** [review]
ok, just two comments. Thanks Marcelo for this PR.

File: Extensions/Cosmos/Cosmos.DataTransfer.CosmosExtension/CosmosExtensionServices.cs (hunk 1 above)
Comment (fonsecamar) [reply Extensions/Cosmos/Cosmos.DataTransfer.CosmosExtension/CosmosExtensionServices.cs:46]: Thanks, @markjbrown.
Readme updated.

File: Extensions/Mongo/Cosmos.DataTransfer.MongoExtension/MongoDataSourceExtension.cs (hunk 2 above)
Comment (fonsecamar) [reply Extensions/Mongo/Cosmos.DataTransfer.MongoExtension/MongoDataSourceExtension.cs:22]: Fixed, @markjbrown .

**Review by markjbrown — APPROVED** [review]

**Comment by fonsecamar** [comment]
[like]  Marcelo Fonseca reacted to your message:
________________________________

**Pull request merged**